### PR TITLE
Add end-to-end emulator test running on CI

### PR DESCRIPTION
**Pull request by stevesoltys**

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -150,10 +157,13 @@ dependencies {
     testRuntimeOnly "org.junit.jupiter:junit-jupiter-engine:$junit5_version"
     testRuntimeOnly "org.junit.vintage:junit-vintage-engine:$junit5_version"
 
+    androidTestImplementation rootProject.ext.aosp_libs
```
Comment (stevesoltys) [inline app/build.gradle:160]: I think I can remove this...

Comment (grote) [reply app/build.gradle:160]: This *should* already get pulled in via `implementation`.

File: app/build.gradle [hunk 2]
```diff
@@ -21,8 +21,15 @@ android {
         minSdk 32 // leave at 32 for robolectric tests
         targetSdk rootProject.ext.targetSdk
         versionNameSuffix "-$gitDescribe"
-        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
+        testInstrumentationRunner "com.stevesoltys.seedvault.KoinInstrumentationTestRunner"
```
Comment (grote) [inline app/build.gradle:24]: Why do we need this? We had instrumentation tests before and they worked, right?

File: app/build.gradle [hunk 3]
```diff
@@ -150,10 +157,13 @@ dependencies {
     testRuntimeOnly "org.junit.jupiter:junit-jupiter-engine:$junit5_version"
     testRuntimeOnly "org.junit.vintage:junit-vintage-engine:$junit5_version"
 
+    androidTestImplementation rootProject.ext.aosp_libs
     androidTestImplementation 'androidx.test:runner:1.4.0'
     androidTestImplementation 'androidx.test:rules:1.4.0'
     androidTestImplementation 'androidx.test.ext:junit:1.1.3'
     androidTestImplementation "io.mockk:mockk-android:$mockk_version"
+
+    androidTestImplementation 'com.kaspersky.android-components:kaspresso:1.5.3'
```
Comment (grote) [inline app/build.gradle]: What is this for? Does this get used anywhere?

File: .idea/runConfigurations/app_emulator.xml [hunk 4]
```diff
@@ -62,7 +62,7 @@
     <option name="SEARCH_ACTIVITY_IN_GLOBAL_SCOPE" value="false" />
     <option name="SKIP_ACTIVITY_VALIDATION" value="false" />
     <method v="2">
-      <option name="RunConfigurationTask" enabled="false" run_configuration_name="seedvault [installEmulatorRelease]" run_configuration_type="GradleRunConfiguration" />
+      <option name="Gradle.BeforeRunTask" enabled="false" tasks="installEmulatorRelease" externalProjectPath="$PROJECT_DIR$/app" vmOptions="" scriptParameters="" />
```
Comment (grote) [inline .idea/runConfigurations/app_emulator.xml:65]: Why is this needed? It wouldn't install the app otherwise?

File: .cirrus.yml [hunk 5]
```diff
@@ -0,0 +1,53 @@
+container:
+  image: ghcr.io/cirruslabs/android-sdk:33
+  kvm: true
+  cpu: 8
+  memory: 16G
+
+check_android_task:
```
Comment (grote) [inline .cirrus.yml]: Maybe we should give this a more speaking name? `Cirrus CI instrumentation tests`?

File: app/src/androidTest/java/com/stevesoltys/seedvault/e2e/BackupRestoreTest.kt [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        verifyCode()
+        chooseBackupLocation()
+
+        val eligiblePackages = launchAllEligibleApps()
+        performBackup(eligiblePackages)
+        uninstallPackages(eligiblePackages)
+        performRestore()
+
+        val packagesAfterRestore = getEligibleApps()
+        assert(eligiblePackages == packagesAfterRestore)
```
Comment (grote) [inline app/src/androidTest/java/com/stevesoltys/seedvault/e2e/BackupRestoreTest.kt]: What apps are installed on the emulator for testing? Do those apps even have data to get backed up? Do we have a way to assert that the data in the backup really matches what we would expect to get restored?

Comment (grote) [reply app/src/androidTest/java/com/stevesoltys/seedvault/e2e/BackupRestoreTest.kt]: Related ticket: https://github.com/seedvault-app/seedvault/issues/548

File: app/build.gradle (hunk 2 above)
Comment (stevesoltys) [reply app/build.gradle:24]: Yeah, this is [to add mockk spies](https://github.com/seedvault-app/seedvault/blob/259ec52054cba4cd5bdbd35832717266829426e6/app/src/androidTest/java/com/stevesoltys/seedvault/KoinInstrumentationTestApp.kt) for some dependencies.

File: app/build.gradle (hunk 3 above)
Comment (stevesoltys) [reply app/build.gradle]: That's [this library](https://github.com/KasperskyLab/Kaspresso) which we use as a [base test class](https://github.com/seedvault-app/seedvault/blob/259ec52054cba4cd5bdbd35832717266829426e6/app/src/androidTest/java/com/stevesoltys/seedvault/e2e/LargeTestBase.kt#L17) and for [some screens](https://github.com/seedvault-app/seedvault/blob/259ec52054cba4cd5bdbd35832717266829426e6/app/src/androidTest/java/com/stevesoltys/seedvault/e2e/screen/impl/RecoveryCodeScreen.kt).

File: .idea/runConfigurations/app_emulator.xml (hunk 4 above)
Comment (stevesoltys) [reply .idea/runConfigurations/app_emulator.xml:65]: I'd have to check again but I recall using regular `adb install` wasn't sufficient as it would remove it from `priv-app` (I think... I didn't investigate much past it not working).

File: app/src/androidTest/java/com/stevesoltys/seedvault/e2e/BackupRestoreTest.kt (hunk 6 above)
Comment (stevesoltys) [reply app/src/androidTest/java/com/stevesoltys/seedvault/e2e/BackupRestoreTest.kt]: Currently [only Etar](https://github.com/seedvault-app/seedvault/blob/259ec52054cba4cd5bdbd35832717266829426e6/.cirrus.yml#L24). Etar will have data because of [this code](https://github.com/seedvault-app/seedvault/blob/259ec52054cba4cd5bdbd35832717266829426e6/app/src/androidTest/java/com/stevesoltys/seedvault/e2e/BackupRestoreTest.kt#L48) which launches every eligible app's activity in an attempt to generate data.

There is no assertion for data just yet but it does validate that the app is re-installed, and the screenshots prove that Etar was successfully backed up and restored - https://cirrus-ci.com/task/5962961425006592

I'm thinking in a subsequent PR we could probably set up a custom AVD `userdata.img` which includes a bunch of apps that have data pre-configured, which we can run this test against and validate data is actually restored (besides viewing the screenshots). Or better yet, maybe we can just restore a Seedvault backup to provision the test emulator instead :wink:

Comment (chirayudesai) [reply app/src/androidTest/java/com/stevesoltys/seedvault/e2e/BackupRestoreTest.kt]: > Or better yet, maybe we can just restore a Seedvault backup to provision the test emulator instead 😉

A reference backup would be handy for other things too!

We can just manually create it from an emulator install, save the backup itself and entire userdata too, and keep updating it as needed.

Comment (stevesoltys) [reply app/src/androidTest/java/com/stevesoltys/seedvault/e2e/BackupRestoreTest.kt]: Created a repository for storing these test backups here:

https://github.com/seedvault-app/seedvault-test-data/releases/tag/1

I will work on another PR to start using these

File: app/build.gradle (hunk 1 above)
Comment (stevesoltys) [reply app/build.gradle:160]: Removed.

File: app/build.gradle (hunk 3 above)
Comment (stevesoltys) [reply app/build.gradle]: Removed this library since we can get by without it and it wasn't providing much value.

File: .cirrus.yml (hunk 5 above)
Comment (stevesoltys) [reply .cirrus.yml]: Done.

File: app/src/androidTest/java/com/stevesoltys/seedvault/e2e/BackupRestoreTest.kt (hunk 6 above)
Comment (stevesoltys) [reply app/src/androidTest/java/com/stevesoltys/seedvault/e2e/BackupRestoreTest.kt]: We are now using the above as a baseline for the E2E test.

The tests will:
1. Provision the emulator with Seedvault.
2. Download and restore the test backup.
3. Run the test afterwards, with a nice set of data and apps.

It will now also record the entire process and upload the recording as an artifact!

Comment (stevesoltys) [reply app/src/androidTest/java/com/stevesoltys/seedvault/e2e/BackupRestoreTest.kt]: @grote as of the latest commit we are now also:

1. Intercepting the backup and restore data, hashing and storing it for each package for both K/V and Full.
2. Asserting that the data we intercepted during restore matches the backup data.
3. Asserting that the backup metadata has no `UNKNOWN_ERROR` cases.

File: app/src/androidTest/java/com/stevesoltys/seedvault/e2e/LargeTestBase.kt [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+        val sharedPreferences = permitDiskReads {
+            PreferenceManager.getDefaultSharedPreferences(targetContext)
+        }
+        sharedPreferences.edit().clear().apply()
+
+        KeyStore.getInstance(ANDROID_KEY_STORE).apply {
+            load(null)
+        }.apply {
+            deleteEntry(KEY_ALIAS_MAIN)
+            deleteEntry(KEY_ALIAS_BACKUP)
```
Comment (grote) [inline app/src/androidTest/java/com/stevesoltys/seedvault/e2e/LargeTestBase.kt:87]: do we need to do all this manually or is there another way like resetting/clearing all app-data?

File: app/build.gradle [hunk 8]
```diff
@@ -210,3 +219,14 @@ tasks.register('installEmulatorRelease', Exec) {
...
     }
 }
+
+tasks.register('clearEmulatorAppData', Exec) {
+    group("emulator")
+
+    doFirst {
+        commandLine "${project.projectDir}/development/scripts/clear_app_data.sh"
+
+        environment "ANDROID_SDK_HOME", android.sdkDirectory.absolutePath
```
Comment (grote) [inline app/build.gradle]: Is this the official variable name? Here, it is called `ANDROID_HOME`.

File: app/development/scripts/clear_app_data.sh [hunk 9]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+
+# assert ANDROID_HOME is set
+if [ -z "$ANDROID_SDK_HOME" ]; then
+  echo "ANDROID_SDK_HOME is not set"
```
Comment (grote) [inline app/development/scripts/clear_app_data.sh]: some variable name question here.

File: app/development/scripts/clear_app_data.sh [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+  exit 1
+fi
+
+SCRIPT_DIR=$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")" &>/dev/null && pwd)
+DEVELOPMENT_DIR=$SCRIPT_DIR/..
+ROOT_PROJECT_DIR=$SCRIPT_DIR/../../..
+
+EMULATOR_DEVICE_NAME=$($ANDROID_SDK_HOME/platform-tools/adb devices | grep emulator | cut -f1)
```
Comment (grote) [inline app/development/scripts/clear_app_data.sh]: this limits execution to emulators. is this intentional? why not allow physical test devices?

File: .idea/runConfigurations/app_emulator.xml (hunk 4 above)
Comment (grote) [reply .idea/runConfigurations/app_emulator.xml:65]: I actually sometimes/often have the issue that Android studio says it installed just fine, but the app wasn't really updated. Maybe this fixes it?

File: app/src/androidTest/java/com/stevesoltys/seedvault/e2e/screen/impl/DocumentPickerScreen.kt [hunk 11]
```diff
@@ -0,0 +1,19 @@
+package com.stevesoltys.seedvault.e2e.screen.impl
+
+import android.widget.EditText
+import com.stevesoltys.seedvault.e2e.screen.UiDeviceScreen
+
+object DocumentPickerScreen : UiDeviceScreen<DocumentPickerScreen>() {
+
+    val createNewFolderButton = findObject { text("CREATE NEW FOLDER") }
+
+    val useThisFolderButton = findObject { text("USE THIS FOLDER") }
```
Comment (grote) [inline app/src/androidTest/java/com/stevesoltys/seedvault/e2e/screen/impl/DocumentPickerScreen.kt:10]: ok these are specific to manage documents permission missing, so it won't work on my test phones which isn't a problem right now, but prevent me from running the tests outside of CI (because I can't run emulators).

File: app/src/androidTest/java/com/stevesoltys/seedvault/e2e/BackupRestoreTest.kt (hunk 6 above)
Comment (grote) [reply app/src/androidTest/java/com/stevesoltys/seedvault/e2e/BackupRestoreTest.kt]: awesome! 

maybe later, we can also check that certain packages in the test data did get backed up (maybe assert the number of packages as well?) and maybe, if test data is fixed, for some packages compare against expected data hash?

Comment (stevesoltys) [reply app/src/androidTest/java/com/stevesoltys/seedvault/e2e/BackupRestoreTest.kt]: Yes definitely! All of those would be great improvements. Test data is static as you said so should work.

File: app/src/androidTest/java/com/stevesoltys/seedvault/e2e/LargeTestBase.kt (hunk 7 above)
Comment (stevesoltys) [reply app/src/androidTest/java/com/stevesoltys/seedvault/e2e/LargeTestBase.kt:87]: The reason I did it this way is because clearing the app data (with adb) seems to close the app, failing the test.

Since we need to clear the state before each test and after restoring the baseline backup, I couldn't think of a better solution. Agree that it's not the best... open to any ideas :smile:

File: app/development/scripts/clear_app_data.sh (hunk 10 above)
Comment (stevesoltys) [reply app/development/scripts/clear_app_data.sh]: The reason I did this was to avoid the person running the tests accidentially targeting their personal phone (like me :laughing:) when running these tests, since it will do some destructive things like uninstalling packages, clearing backups, wiping seedvault, etc.

I can't test on a physical device unfortunately, but if we could think of away around the above could avoid this I'm sure?

File: .idea/runConfigurations/app_emulator.xml (hunk 4 above)
Comment (stevesoltys) [reply .idea/runConfigurations/app_emulator.xml:65]: Could be yeah! I haven't tested on a real device since I only have one, but it's worth trying!

File: app/src/androidTest/java/com/stevesoltys/seedvault/e2e/screen/impl/DocumentPickerScreen.kt (hunk 11 above)
Comment (stevesoltys) [reply app/src/androidTest/java/com/stevesoltys/seedvault/e2e/screen/impl/DocumentPickerScreen.kt:10]: Ahh yeah. Sorry about that. One option is if we can somehow get the [generic system images](https://developer.android.com/topic/generic-system-image/releases) to boot in an emulator, we can test with the permission. I couldn't get this working though :disappointed:. If you know how let me know and we can switch over, it would be better without all the Google apps anyways.

File: app/build.gradle (hunk 8 above)
Comment (stevesoltys) [reply app/build.gradle]: Switched!

File: app/development/scripts/clear_app_data.sh (hunk 10 above)
Comment (grote) [reply app/development/scripts/clear_app_data.sh]: We already have some instrumentation tests like this. Do these tests even work on personal phones? Doesn't it require the platform signing key?

File: app/src/androidTest/java/com/stevesoltys/seedvault/e2e/screen/impl/DocumentPickerScreen.kt (hunk 11 above)
Comment (grote) [reply app/src/androidTest/java/com/stevesoltys/seedvault/e2e/screen/impl/DocumentPickerScreen.kt:10]: What we could do instead is *not* request the MANAGE_DOCUMENTS permission for instrumentation test (via a modified manifest). So every device would behave the same.

**Review by grote — APPROVED** [review]

File: app/development/scripts/clear_app_data.sh (hunk 10 above)
Comment (stevesoltys) [reply app/development/scripts/clear_app_data.sh]: It does, but for instance this script in particular could be run against a personal device and clear the Seedvault data. Maybe just need to enforce only one device is connected?

Comment (grote) [reply app/development/scripts/clear_app_data.sh]: @chirayudesai is there a way to check for a userdebug build device? Maybe that would be more useful than single device?

Comment (chirayudesai) [reply app/development/scripts/clear_app_data.sh]: `getprop ro.build.type` - user, userdebug, or eng

Also, `ro.build.tags`. `test-keys` for dev, `release-keys` for prod (typically)

Ref: https://dumps.tadiphone.dev/dumps/google/lynx/-/blob/lynx-user-13-TQ3A.230605.012-10204971-release-keys/system/system/build.prop?ref_type=heads
